pkg/scheduler/queue/user_request_queue.go

```diff
@@ -0,0 +1,92 @@
+package queue
+
+import (
+	"strconv"
+
+	"github.com/prometheus/client_golang/prometheus"
+
+	"github.com/cortexproject/cortex/pkg/util"
+)
+
+type userRequestQueue interface {
+	enqueueRequest(Request)
+	dequeueRequest(int64, bool) Request
+	length() int
+}
+
+type FIFORequestQueue struct {
+	queue       chan Request
+	userID      string
+	queueLength *prometheus.GaugeVec
+}
+
+func NewFIFORequestQueue(queue chan Request, userID string, queueLength *prometheus.GaugeVec) *FIFORequestQueue {
+	return &FIFORequestQueue{queue: queue, userID: userID, queueLength: queueLength}
+}
+
+func (f *FIFORequestQueue) enqueueRequest(r Request) {
+	f.queue <- r
+	if f.queueLength != nil {
+		f.queueLength.With(prometheus.Labels{
+			"user":     f.userID,
+			"priority": strconv.FormatInt(r.Priority(), 10),
+			"type":     "fifo",
+		}).Inc()
+	}
+}
+
+func (f *FIFORequestQueue) dequeueRequest(_ int64, _ bool) Request {
+	r := <-f.queue
+	if f.queueLength != nil {
+		f.queueLength.With(prometheus.Labels{
+			"user":     f.userID,
+			"priority": strconv.FormatInt(r.Priority(), 10),
+			"type":     "fifo",
+		}).Dec()
+	}
+	return r
+}
+
+func (f *FIFORequestQueue) length() int {
+	return len(f.queue)
+}
+
+type PriorityRequestQueue struct {
+	queue       *util.PriorityQueue
+	userID      string
+	queueLength *prometheus.GaugeVec
+}
+
+func NewPriorityRequestQueue(queue *util.PriorityQueue, userID string, queueLength *prometheus.GaugeVec) *PriorityRequestQueue {
+	return &PriorityRequestQueue{queue: queue, userID: userID, queueLength: queueLength}
+}
+
+func (f *PriorityRequestQueue) enqueueRequest(r Request) {
+	f.queue.Enqueue(r)
+	if f.queueLength != nil {
+		f.queueLength.With(prometheus.Labels{
+			"user":     f.userID,
+			"priority": strconv.FormatInt(r.Priority(), 10),
+			"type":     "priority",
+		}).Inc()
```

Comment:
We can just use 

```
f.queueLength.WithLabelValues(f.userID, strconv.FormatInt(r.Priority(), 10), "priority").Inc()
```

Reply:
I was just worried about too many labels, but if your ok I'm good too :) Addressed in https://github.com/cortexproject/cortex/pull/5605/commits/96ff019ddba591815be280b4f9c9aacf63733a42